mempool/cat/cache.go

```diff
@@ -0,0 +1,291 @@
+package cat
+
+import (
+	"container/list"
+	"time"
+
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+	"github.com/tendermint/tendermint/types"
+)
+
+// LRUTxCache maintains a thread-safe LRU cache of raw transactions. The cache
+// only stores the hash of the raw transaction.
+// NOTE: This has been copied from mempool/cache with the main diffence of using
+// tx keys instead of raw transactions.
+type LRUTxCache struct {
+	staticSize int
+
+	mtx      tmsync.Mutex
+	cacheMap map[types.TxKey]*list.Element
+	list     *list.List
+}
+
+func NewLRUTxCache(cacheSize int) *LRUTxCache {
+	return &LRUTxCache{
+		staticSize: cacheSize,
+		cacheMap:   make(map[types.TxKey]*list.Element, cacheSize),
+		list:       list.New(),
+	}
+}
+
+func (c *LRUTxCache) Reset() {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	c.cacheMap = make(map[types.TxKey]*list.Element, c.staticSize)
+	c.list.Init()
+}
+
+func (c *LRUTxCache) Push(txKey types.TxKey) bool {
+	if c.staticSize == 0 {
+		return true
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	moved, ok := c.cacheMap[txKey]
+	if ok {
+		c.list.MoveToBack(moved)
+		return false
+	}
+
+	if c.list.Len() >= c.staticSize {
+		front := c.list.Front()
+		if front != nil {
+			frontKey := front.Value.(types.TxKey)
+			delete(c.cacheMap, frontKey)
+			c.list.Remove(front)
+		}
+	}
+
+	e := c.list.PushBack(txKey)
+	c.cacheMap[txKey] = e
+
+	return true
+}
+
+func (c *LRUTxCache) Remove(txKey types.TxKey) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	e := c.cacheMap[txKey]
+	delete(c.cacheMap, txKey)
+
+	if e != nil {
+		c.list.Remove(e)
+	}
+}
+
+func (c *LRUTxCache) Has(txKey types.TxKey) bool {
+	if c.staticSize == 0 {
+		return false
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	_, ok := c.cacheMap[txKey]
+	return ok
+}
+
+type EvictedTxInfo struct {
+	timeEvicted time.Time
+	priority    int64
+	gasWanted   int64
+	sender      string
+	size        int64
+}
+
+type EvictedTxCache struct {
+	staticSize int
+
+	mtx   tmsync.Mutex
+	cache map[types.TxKey]*EvictedTxInfo
+}
+
+func NewEvictedTxCache(size int) *EvictedTxCache {
+	return &EvictedTxCache{
+		staticSize: size,
+		cache:      make(map[types.TxKey]*EvictedTxInfo),
```

Comment:
Can you please explain why don't we allocate the cache size according to the supplied `size`? similar to what is done for the `LRUTxCache`.

Reply:
> We could also do the same thing as the other cache and add a list here to avoid needing to run a for loop over all items whenever we need to evict but this is not on the hot path so I don't think the tradeoff makes sense.

If the cache becomes full, then this linear search i.e., the loop takes place for every push operation, and the performance degradation MAY be noticeable. However, it depends on  how frequent elements are pushed to the `EvictedTxCache`.

Reply:
Btw, I am not suggesting to implement the cache optimization in this PR :) , just something to have in mind for the future.

Reply:
Yeah it seemed apparent from the initial design of the priority mempool that eviction performance wasn't an important characteristic. For example, you may notice that for eviction, the mempool also perform a `sort.Slice` on *ALL* txs currently in the mempool to know which transactions have the lowest priority and should therefore be evicted. 

We have metrics tracking the number of evicted transactions so we can track how frequent this occurs and if it becomes a problem in the future, then we can modify these parts

Reply:
> We have metrics tracking the number of evicted transactions so we can track how frequent this occurs and if it becomes a problem in the future, then we can modify these parts

That would be helpful.

Reply:
I'd suggest creating a GH issue (if we don't have one already) to keep track of any performance improvements we come across for our future reference.